drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2129 @@
+{
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	if (dev == NULL || payload == NULL) {
+		return -EINVAL;
+	}
+
+	if (payload->ccc.id == I3C_CCC_ENTDAA) {
+		return -EINVAL;
+	}
+
+	/* Check if payload has targets when sending a direct CCC */
+	if (!i3c_ccc_is_payload_broadcast(payload) &&
+	    (payload->targets.payloads == NULL || payload->targets.num_targets == 0)) {
+		return -EINVAL;
+	}
+
+	if (payload->ccc.data_len > 0 && payload->ccc.data == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_get(dev);
+#endif
+	/* Prevent the clocks to be stopped during the transaction */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	/* Mark current transfer as CCC */
+	data->msg_state = STM32_I3C_MSG_CCC;
+	data->ccc_payload = payload;
+	data->ccc_target_idx = 0;
+	data->ccc_target_payload = payload->targets.payloads;
+
+	payload->ccc.num_xfer = 0;
+
+	for (size_t i = 0; i < payload->targets.num_targets; i++) {
+		payload->targets.payloads->num_xfer = 0;
+	}
+
+	/* Start CCC transfer */
+	LL_I3C_ControllerHandleCCC(i3c, payload->ccc.id, payload->ccc.data_len,
+				   (i3c_ccc_is_payload_broadcast(payload)
+					    ? LL_I3C_GENERATE_STOP
+					    : LL_I3C_GENERATE_RESTART));
+
+	/* Wait for CCC to complete */
+	if (k_sem_take(&data->device_sync_sem, STM32_I3C_TRANSFER_TIMEOUT) != 0) {
+		i3c_stm32_clear_err(dev, false);
+		return -ETIMEDOUT;
+	}
+
+	if (data->msg_state == STM32_I3C_MSG_ERR) {
+		i3c_stm32_clear_err(dev, false);
+		return -EIO;
+	}
+
+	k_mutex_unlock(&data->bus_mutex);
+
+	return 0;
+}
+
+/* Handles the ENTDAA CCC */
+static int i3c_stm32_do_daa(const struct device *dev)
+{
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_get(dev);
+#endif
+	/* Prevent the clocks to be stopped during the transaction */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	/* Mark current transfer as DAA */
+	data->msg_state = STM32_I3C_MSG_DAA;
+
+	/* Disable TXFNF interrupt, the RXFNE interrupt will enable it once all PID bytes are
+	 * received
+	 */
+	LL_I3C_DisableIT_TXFNF(i3c);
+
+	/* Start DAA */
+	LL_I3C_ControllerHandleCCC(i3c, I3C_CCC_ENTDAA, 0, LL_I3C_GENERATE_STOP);
+
+	/* Wait for DAA to finish */
+	if (k_sem_take(&data->device_sync_sem, STM32_I3C_TRANSFER_TIMEOUT) != 0) {
+		return -ETIMEDOUT;
+	}
+
+	if (data->msg_state == STM32_I3C_MSG_ERR) {
+		i3c_stm32_clear_err(dev, false);
+		/* Enable TXFNF interrupt in case an error occurred before it was enabled by RXFNE
+		 */
+		LL_I3C_EnableIT_TXFNF(i3c);
+		return -EIO;
+	}
+
+	k_mutex_unlock(&data->bus_mutex);
+
+	return 0;
+}
```

Comment:
How is this to gather what it found with ENTDAA and the values it assigned?

Reply:
Are you asking how to retrieve data like the PID from the target? If so, you can refer to the logic implemented in the ISR, particularly in the function i3c_stm32_event_isr_rx.